### PR TITLE
Feature/upload

### PR DESCRIPTION
**Pull request by liujoshua**
Queue pending uploads by saving them locally for future processing

### CODE REVIEW

**Comment by dephillipsmichael** [comment]
Unit tests need updated to reflect new queue/upload flow.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 1]
```diff
@@ -26,51 +28,145 @@
...
 
     private final ForConsentedUsersApi api;
     private final StudyUploadEncryptor encryptor;
-    private final OkHttpClient s3OkhttpClient;
+    private final UploadDAO uploadDAO;
+    private final OkHttpClient okHttpClient = new OkHttpClient.Builder()
+            .connectTimeout(30, TimeUnit.SECONDS)
+            .readTimeout(30, TimeUnit.SECONDS)
+            .writeTimeout(30, TimeUnit.SECONDS)
+            .retryOnConnectionFailure(false).build();
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java:78]: It seems like this would be a nice way to get free retry logic.  Did you consider enabling it?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 2]
```diff
@@ -26,51 +28,145 @@
...
+                });
     }
 
-    @NonNull
-    Single<UploadValidationStatus> uploadToS3(UploadFile uploadFile, UploadSession session) {
+    // determines whether an upload failure is permanent, i.e. no point in retrying
+    boolean isPermanentFailure(UploadFile uploadFile, Throwable throwable) {
+        if (throwable instanceof IllegalStateException) {
+            // some pre-condition failed
+            return true;
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: Can you elaborate more in comments about specifically what would cause an IllegalStateException here?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 3]
```diff
@@ -26,51 +28,145 @@
...
+            return Single.just(null);
+        }
 
-        return Single.zip(uploadFileSingle, uploadSessionSingle, this::uploadToS3).flatMap(i -> i);
+        return getUploadSession(uploadFile)
+                .flatMap(uploadSession -> getUploadValidationStatus(uploadSession.getId()))
+                .flatMap((uploadValidationStatus) -> {
+                    if (uploadValidationStatus == null) {
+                        LOG.info("Could not find uploadValidationStatus for filename: " + filename);
+                        return Single.just(null);
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: Should this be an error message of some sort?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 4]
```diff
@@ -26,51 +28,145 @@
...
+                        uploadDAO.removeUploadAndSession(filename);
+                        return Single.just(null);
+                    } else if (UploadStatus.VALIDATION_FAILED == uploadValidationStatus.getStatus()) {
+                        LOG.info("Cleaning up validation-failed upload with filename: " + filename
+                                + ", id: " + uploadValidationStatus.getId());
+                        uploadDAO.removeUploadAndSession(filename);
+                        return Single.just(null);
+                    }
+
+                    return getUploadSession(uploadFile)
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: This return is crazy complex.  Can you add comments on why you would do the entire upload flow again here?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 5]
```diff
@@ -26,51 +28,145 @@
...
+                            .flatMapCompletable(uploadSession -> uploadToS3(uploadFile, uploadSession))
+                            .andThen(getUploadValidationStatus((uploadValidationStatus.getId())));
+                }).doOnSuccess(uploadValidationStatus -> {
+                    LOG.debug("uploadToS3 succeded, uploadValidationStatus: " + uploadValidationStatus);
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: Wouldn't this make more sense to Log in the block above?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 6]
```diff
@@ -104,24 +203,28 @@ public UploadManager(AuthenticationManager authenticationManager, StudyUploadEnc
...
-                    // TODO: insert upload session to DB, link to file metadata
-                    LOG.info("Received upload session with id: " + uploadSession.getId());
+                    LOG.info("Received uploadToS3 session with id: " + uploadSession.getId());
+                    uploadDAO.putUploadSession(uploadFile.filename, uploadSession);
                 });
     }
 
-    UploadFile persist(String filename, Archive archive) throws NoSuchAlgorithmException, IOException {
+    @WorkerThread
+    UploadFile persist(String filename, Archive archive) throws IOException,
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: Can you put method headers and descriptions on all the methods in this class?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/BridgeManagerProvider.java [hunk 7]
```diff
@@ -113,6 +117,8 @@ private BridgeManagerProvider(@NonNull Context applicationContext) {
     @NonNull
     private final AccountDAO accountDAO;
     @NonNull
+    private final UploadDAO uploadDAO;
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/BridgeManagerProvider.java]: How/when does the uploadDAO get cleared out after sign out?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/util/retrofit/RxUtils.java [hunk 8]
```diff
@@ -18,7 +19,8 @@
     public static <T> Single<T> toBodySingle(@NonNull Call<T> call) {
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/util/retrofit/RxUtils.java:27]: Can you add a method header describing this?

File: researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/TaskHelper.java [hunk 9]
```diff
@@ -189,19 +188,25 @@ void uploadTaskResult(TaskResult taskResult, Archive.WithBridgeConfig withBridge
...
 
-        // At this point, the upload request has been processed and saved,
-        // so it is safe to delete the temporary data logger files
-        for (Result result : results) {
-            if (result instanceof FileResult) {
-                FileResult fileResult = (FileResult) result;
-                DataLoggerManager.getInstance().deleteFileStatus(fileResult.getFile());
-            }
-        }
+                    bridgeManagerProvider.getUploadManager().uploadToS3().toCompletable().await();
```
Comment (dephillipsmichael) [inline researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/TaskHelper.java]: This would imply that this method uploadTaskResult, needs to be off the main thread correct?  Can you add a method header/description/annotation to reflect this?

Also, if this is on a different thread, why not just do ".await()" for the queue upload as well?

File: researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/TaskHelper.java [hunk 10]
```diff
@@ -189,19 +188,25 @@ void uploadTaskResult(TaskResult taskResult, Archive.WithBridgeConfig withBridge
...
+                .queueUpload(archiveFilename, bridgeArchiveBuilder.build())
+                .doOnSuccess(uploadFile -> {
+                    logger.debug("Calling uploadToS3() to upload pending uploads");
+
+                    // At this point, the upload request has been processed and saved,
+                    // so it is safe to delete the temporary data logger files
+                    for (Result result : results) {
+                        if (result instanceof FileResult) {
+                            FileResult fileResult = (FileResult) result;
+                            DataLoggerManager.getInstance().deleteFileStatus(fileResult.getFile());
```
Comment (dephillipsmichael) [inline researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/TaskHelper.java]: I'm not entirely sure if deleteFileStatus is thread-safe.  I didn't intend on it to be called off the  main thread.

**Comment by dephillipsmichael** [comment]
IMO, it's hard to follow the logic of the UploadManager, which is making it difficult to truly review.  The iOS version, SBBUploadManager, is much easier to understand since it has detailed documentation, and specifically deals with all success / http error code responses.  I'd suggest adding detailed comments to your code before I can approve, using iOS as a reference.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/util/retrofit/RxUtils.java (hunk 8 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/util/retrofit/RxUtils.java:27]: Added

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/BridgeManagerProvider.java (hunk 7 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/BridgeManagerProvider.java]: It currently doesn't. I created a separate story/sub-task in JIRA for this

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java (hunk 1 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java:78]: Good idea. I have a branch that moves the call to the network part of the upload into a SyncProvider, which I believe handles connection failures. I'll test it with this flag

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/BridgeManagerProvider.java (hunk 7 above)
Comment (dephillipsmichael) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/BridgeManagerProvider.java]: I think that should be a part of this story.  Seems like a really easy task, no?  On successful sign out, call clear.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 11]
```diff
@@ -26,76 +30,257 @@
...
-                .fromCallable(() -> persist(filename, archive)).cache();
+    public Single<UploadFile> queueUpload(String filename, Archive archive) {
+        return Single.fromCallable(() -> persist(filename, archive))
+                .subscribeOn(Schedulers.io());
+    }
 
-        Single<UploadSession> uploadSessionSingle = uploadFileSingle
-                .flatMap(this::requestUploadSession);
+    /**
+     * This triggers an attempt to upload all the previously queued upload files
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: How many threads does the Schedulers.io() have in it's thread pool?  This might be important to note in the comments.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 12]
```diff
@@ -26,76 +30,257 @@
...
+     * You must call queueUpload before this method will have anything to operate on
+     *
+     * @return Observable with information on if the upload was successful or not
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: "Observable with information on if the upload was successful or not"... which upload?  all of them?  Seems like this might make sense as a Void return type if you can't return the status of all of them as a list.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 13]
```diff
@@ -26,76 +30,257 @@
...
+
+        return filenameObservable
+                .map(filename -> uploadDAO.getUploadFile(filename))
+                .flatMap(uploadFile -> processUploadFile(uploadFile).toObservable())
+                .toCompletable();
+    }
 
-        return Single.zip(uploadFileSingle, uploadSessionSingle, this::uploadToS3).flatMap(i -> i);
+    /**
+     * Performs next upload step for a file.
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: Honestly, this doesn't really help me understand what is going on in the method.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 14]
```diff
@@ -26,76 +30,257 @@
...
+import rx.Observable;
 import rx.Single;
+import rx.schedulers.Schedulers;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
 
 /**
  * Created by jyliu on 1/30/2017.
  */
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: I see a lot of references to "the next step in the upload process" so i think it would be good to give an overview of the full upload process here.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 15]
```diff
@@ -26,76 +30,257 @@
...
+                sessionSingle,
+                statusSingle,
+                this::processUploadForValidationStatus
+        ).flatMapCompletable(i -> i);
     }
 
+    /**
+     * Performs the next upload step, according to the UploadValidationStatus. The normal
+     * transition of UploadValidationStatus is from REQUESTED to VALIDATION_IN_PROGRESS to
+     * REQUESTED.
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: do you mean to COMPLETE at the end?

Comment (dephillipsmichael) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: sorry i mean SUCCESS

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 16]
```diff
@@ -26,76 +30,257 @@
...
+                return uploadToS3(uploadFile, uploadSession);
+            case SUCCEEDED:
+            case DUPLICATE:
+                return dequeueUpload(uploadFile.filename);
+            case VALIDATION_IN_PROGRESS:
+                LOG.debug("Validation in progress for filename: " + uploadFile.filename +
+                        ", uploadId" + uploadSession.getId());
+                break;
+            case VALIDATION_FAILED:
+                LOG.debug("Validation failed for filename: " + uploadFile.filename +
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java:242]: Should this dequeueUpload as well?  Or will this fail silently and it will be attempted again?

Comment (dephillipsmichael) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java:242]: Never mind, the comment at the end answer my questions, nice!

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 17]
```diff
@@ -26,76 +30,257 @@
...
+    Completable dequeueUpload(@NonNull String filename) {
+        checkNotNull(filename, "filename required");
+
+        if (getFile(filename).delete()) {
+            LOG.warn("Successfully deleted upload file: " + filename + ", "
+                    + "removing upload from queue");
+            return Completable.fromAction(() -> uploadDAO.removeUploadAndSession(filename));
+        } else {
+            LOG.warn("Failed to delete upload file: " + filename);
+            return Completable.error(new IOException("Failed to delete upload file: " + filename));
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: I usually like to make delete file errors log but not actually throw an error like the completable error, because usually they mean the file didn't exists to delete and that is like a success

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 18]
```diff
@@ -26,76 +30,257 @@
...
+     */
+    @NonNull
+    Single<UploadValidationStatus> getUploadValidationStatus(@NonNull String uploadId) {
+        checkNotNull(uploadId, "uploadId required");
+
+        return RxUtils.toBodySingle(api.getUploadStatus(uploadId));
+    }
+
+    @NonNull
+    Completable uploadToS3(UploadFile uploadFile, UploadSession session) {
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: method header?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 19]
```diff
@@ -26,76 +30,257 @@
...
         File file = getFile(uploadFile.filename);
+        checkArgument(file.exists(), "Non-existent file: " + file.getAbsolutePath());
+
+        Single<UploadSession> sessionSingle = Single.just(session)
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java:300]: What is going on here?  You take an UploadSession, and wrap it in a Single and then operate on that same variable?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java [hunk 20]
```diff
@@ -26,76 +30,257 @@
...
+                getS3Service(freshSession)
+                        .uploadToS3(
+                                freshSession.getUrl(),
+                                requestBody,
+                                uploadFile.md5Hash,
+                                uploadFile.contentType)))
+                .doOnSuccess(aVoid -> {
+                    LOG.info("S3 upload succeeded for id: " + session.getId());
+
+                    // call upload complete on a computation thread
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java:325]: Seems odd you chose this for a computation thread when all the other network calls happen on IO.  What was your reasoning?

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/dao/UploadDAO.java [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.Set;
+
+/**
+ * Created by jyliu on 3/10/2017.
+ */
+
+public class UploadDAO extends SharedPreferencesJsonDAO {
+    private static final Logger logger = LoggerFactory.getLogger(UploadDAO.class);
+
+    private static final String PREFERENCES_FILE  = "uploads";
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/dao/UploadDAO.java:21]: The more I think about it, we should move all the DAOs form shared preferences to StorageAccess.  That way we can just delete all the files in StorageAccess' FileAccess when the user signs out and it will make life a lot easier.

File: android-sdk/src/test/java/org/sagebionetworks/bridge/android/manager/UploadManagerTest.java [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+
+    private UploadManager spyUploadManager;
+
+    private UploadManager.UploadFile uploadFile;
+    private UploadSession uploadSession;
+    private UploadValidationStatus uploadValidationStatus;
+
+
+    @Before
+    public void setupTest() {
```
Comment (dephillipsmichael) [inline android-sdk/src/test/java/org/sagebionetworks/bridge/android/manager/UploadManagerTest.java:69]: Nice test coverage on this file

File: researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/TaskHelper.java [hunk 23]
```diff
@@ -195,16 +198,25 @@ void uploadTaskResult(TaskResult taskResult, Archive.WithBridgeConfig withBridge
...
-        for (Result result : results) {
-            if (result instanceof FileResult) {
-                FileResult fileResult = (FileResult) result;
-                DataLoggerManager.getInstance().deleteFileStatus(fileResult.getFile());
-            }
-        }
+                .queueUpload(archiveFilename, bridgeArchiveBuilder.build())
+                .doOnSuccess(uploadFile -> {
+                    logger.debug("Attempting upload in io() thread");
+                    bridgeManagerProvider.getUploadManager()
```
Comment (dephillipsmichael) [inline researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/TaskHelper.java]: If upload needs to happen on the IO thread, I think the upload manager should put it on the I/O thread, since it shouldn't be the responsibility of the caller to know to call process upload file on a specific thread.

**Comment by dephillipsmichael** [comment]
The code is a lot easier to follow now, thank you for adding all the documentation and re-organizing so that is more readable.  I still have some issues from my comments, but I was able to understand everything fully.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java (hunk 11 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: These schedulers are optimized/managed for us by the platform/library.

The comment on Schedulers.io() indicate it's backed by an Executor thread-pool that will grow as needed.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java (hunk 14 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: Good point

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/BridgeManagerProvider.java (hunk 7 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/BridgeManagerProvider.java]: Yea that makes sense, I'll clear all for now, but that is not the full story. When the user signs out, or a different user signs in, files waiting for upload and already have a UploadSession (pre-signed URL) should still get uploaded.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java (hunk 12 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: I did change it to an Observable<Void>, but I changed it back to Completable because Observable<Void> didn't provide an advantage (can't do anything with it). I did some googling on Observable<Void>, Completables, and Observable<Completable> and found a more readable and useful way to run a collection of Completables: Completable.mergeDelayError

(see https://github.com/ReactiveX/RxJava/issues/3807)

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java (hunk 17 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java]: Yea that sounds better

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java (hunk 20 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java:325]: Fixed, you're right I didn't mean to do that. TUrns out, my attempt to run it on computation thread wouldn't have worked, because RxUtils.toBodySingle already called subscribeOn(Schedulers.io())

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java (hunk 19 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java:300]: If the session is not expired or nearing expiry, yes, it is the identity function, we just reuse the Single<Session> created from the value we were passed as a param. Otherwise, the Single<Session> is the result of a request to Bridge for a new Session.

The Single<Session> (containing either the Session in param, or new one for Bridge) is then used for the S3Upload.

Do you think it would be easier to test/understand if i split it up?

So instead of uploadToS3(UplaodFile,UploadSession) that refreshes session and uploads, two separate methods, something like

refreshStaleSession(UploadSession session)

uploadToS3(UploadFile file, String url)

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/dao/UploadDAO.java (hunk 21 above)
Comment (liujoshua) [reply android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/dao/UploadDAO.java:21]: Storage Access is in RS, android-sdk has no RS dependency. You are on the right track that researchstack-sdk should implement StorageAccess or otherwise hook in the DAOs.

File: researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/TaskHelper.java (hunk 23 above)
Comment (liujoshua) [reply researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/TaskHelper.java]: I agree, I believe this line isn't needed. UploadManager is already handling the threading.

File: android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/AuthenticationManager.java [hunk 24]
```diff
@@ -193,11 +205,25 @@ public Completable resendEmailVerification(@NonNull final String email) {
...
+        }
+
         return RxUtils.toBodySingle(authenticationApi.signOut())
                 .doOnSuccess(message -> {
                     accountDAO.setSignIn(null);
                     accountDAO.setUserSessionInfo(null);
                     accountDAO.setStudyParticipant(null);
+
+                    for (AuthenticationEventListener listener : listeners) {
+                        listener.onSignedOut(email);
```
Comment (dephillipsmichael) [inline android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/AuthenticationManager.java:225]: It's usually not the best practice to call listeners on a background thread.  Can we post these to the main thread?
